contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,519 @@
+        bytes message;
+    }
+}
+
+impl Message {
+    pub fn abi_encode(self, env: &Env) -> Bytes {
+        let msg = match self {
+            Self::InterchainTransfer(types::InterchainTransfer {
+                token_id,
+                source_address,
+                destination_address,
+                amount,
+                data,
+            }) => InterchainTransfer {
+                messageType: U256::from(types::MessageType::InterchainTransfer as u32),
+                tokenId: FixedBytes::<32>::new(token_id.into()),
+                sourceAddress: source_address.to_alloc_vec().into(),
+                destinationAddress: destination_address.to_alloc_vec().into(),
+                amount: amount.try_into().unwrap_or_default(),
+                data: data.map(|d| d.to_alloc_vec()).unwrap_or_default().into(),
+            }
+            .abi_encode_params(),
+            Self::DeployInterchainToken(types::DeployInterchainToken {
+                token_id,
+                name,
+                symbol,
+                decimals,
+                minter,
+            }) => DeployInterchainToken {
+                messageType: U256::from(types::MessageType::DeployInterchainToken as u32),
+                tokenId: FixedBytes::<32>::new(token_id.into()),
+                name: soroban_string_to_string(name),
+                symbol: soroban_string_to_string(symbol),
+                decimals: decimals.try_into().unwrap_or_default(),
+                minter: minter.map(|m| m.to_alloc_vec()).unwrap_or_default().into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
+    }
+
+    pub fn abi_decode(env: &Env, payload: &Bytes) -> Result<Self, MessageError> {
+        ensure!(payload.len() >= 32, MessageError::InsufficientMessageLength);
+
+        let binding = payload.to_alloc_vec();
+        let payload_array = binding.as_slice();
+
+        let message_type = match MessageType::abi_decode(&payload_array[0..32], true) {
+            Ok(value) => value,
+            Err(_) => return Err(MessageError::InvalidMessageType),
+        };
+
+        let message = match message_type {
+            MessageType::InterchainTransfer => {
+                let decoded = match InterchainTransfer::abi_decode_params(payload_array, true) {
+                    Ok(value) => value,
+                    Err(_) => return Err(MessageError::AbiDecodeFailed),
+                };
+
+                Self::InterchainTransfer(types::InterchainTransfer {
+                    token_id: BytesN::from_array(env, &decoded.tokenId.into()),
+                    source_address: Bytes::from_slice(env, decoded.sourceAddress.as_ref()),
+                    destination_address: Bytes::from_slice(
+                        env,
+                        decoded.destinationAddress.as_ref(),
+                    ),
+                    amount: convert_to_i128(decoded.amount)?,
+                    data: match decoded.data.len() {
```

Comment:
Could this be written as a one-liner?
```suggestion
                    data: (!decoded.data.is_empty()).then(|| Bytes::from_slice(env, decoded.data.as_ref())),
```